### PR TITLE
spectral extraction: live-preview and export of background spectrum

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request adds live-visualization and export (UI and API) support for the 1D spectrum representing the background region within the spectral extraction plugin.

~NOTE: this requires https://github.com/astropy/specreduce/pull/143 (until then, the new test coverage will fail)~

[Rendered updated plugin docs](https://jdaviz--1682.org.readthedocs.build/en/1682/specviz2d/plugins.html#background)

https://user-images.githubusercontent.com/877591/192842956-2fba147d-d027-48e7-96e2-fc0f75d7cb77.mov

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

New Features > Specviz2d:

Spectral extraction plugin now supports visualizing and exporting the 1D spectrum associated with the background region.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by camipacifici** [comment]
How about a thicker solid line for the extracted spectrum and a thin solid line for the background spectrum?

**Comment by kecnry** [comment]
@camipacifici @PatrickOgle - I actually like the thinner line idea, what do you think?  (Here are videos for the dotted and thin styling and how they both react to zooming)


https://user-images.githubusercontent.com/877591/194934041-29280ff9-0353-42e1-9415-979df0d3f64c.mov

https://user-images.githubusercontent.com/877591/194934052-d1ed4f6b-93d4-48c1-bc60-047c126d1f74.mov

**Comment by camipacifici** [comment]
I like the thin line too!

File: docs/specviz2d/plugins.rst [hunk 1]
```diff
@@ -94,9 +94,11 @@ The background step of the plugin allows for creating background and background-
 images via `specreduce.background <https://specreduce.readthedocs.io/en/latest/#module-specreduce.background>`_.
 
 Once you interact with any of the inputs in the background step or hover over that area
-of the plugin, the live visualization will change to show the center (dotted line) and edges
-(solid lines) of the background region(s).  Choose between creating the background
-around the trace defined in the Trace section, or around a "Manual" flat trace.
+of the plugin, the live visualization in the 2d spectrum viewer will change to show the center 
+(dotted line) and edges (solid lines) of the background region(s).  The 1D representation of the
+background will also be visualized in the 1D spectrum viewer (thin solid line).
```
Comment (ojustino) [inline docs/specviz2d/plugins.rst:99]:
```suggestion
of the plugin, the live visualization in the 2D spectrum viewer will change to show the center 
(dotted line) and edges (solid lines) of the background region(s).  The 1D representation of the
background will also be visualized in the 1D spectrum viewer (thin, solid line).
```

File: docs/specviz2d/plugins.rst [hunk 2]
```diff
@@ -94,9 +94,11 @@ The background step of the plugin allows for creating background and background-
...
+
+Choose between creating the background around the trace defined in the Trace section, or around a "Manual" flat trace.
```
Comment (ojustino) [inline docs/specviz2d/plugins.rst:101]: I found this sentence a little hard to parse. What about something like this? "Backgrounds can either be created around the trace defined in the earlier Trace section or around a new, flat trace by selecting a "Manual" Background type."

**Review by ojustino — COMMENTED** [review]
The code changes seem reasonable to me. A few observations:

1. I am fine with the solid line, too. However, depending on the placement of the initial background region(s), it may not be noticeable if you aren't already looking for it. How do you feel about giving it a different color?

2. When I choose the "Manual" option, I don't see a preview line and the 2D viewer takes 2-3 seconds to adjust when I change the center pixel. (The lag is also present in `main`.) Exporting a spectrum having chosen "Manual" brings up the "background regions overlapped" error from specreduce. "TwoSided" and "OneSided" are snappy and don't raise any errors for me.

3. Sometimes when I export the background 1D spectrum and then scroll up to change its appearance in Plot Options, the preview line persists in the 1D viewer:

image

The preview disappears if I scroll back into the Trace or Extraction sections of the Spectral Extraction plugin, so it looks like the trigger is leaving the Spectral Extraction plugin without registering a mouse move (e.g., pressing "Extract" and immediately scrolling up). I'm not sure how many would run into this but thought I'd mention it.

**Comment by kecnry** [comment]
> How do you feel about giving it a different color?

I would like to avoid this, if possible.  We use the same color for all "temporary" previews from plugins at the moment.

> When I choose the "Manual" option, I don't see a preview line and the 2D viewer takes 2-3 seconds to adjust when I change the center pixel.

I've come across this a few times as well, but is unrelated to this PR (feel free to open an issue if one doesn't already exist).

> Sometimes when I export the background 1D spectrum and then scroll up to change its appearance in Plot Options, the preview line persists in the 1D viewer

I suspect the mousemove isn't triggering a change in the "section", but so long as the plugin is opened, I would expect the previews to stay visible, in general.  If you moused over a different "section" of the plugin and that wasn't registered, then that may be a bug, although again unrelated to this PR.

File: docs/specviz2d/plugins.rst (hunk 2 above)
Comment (kecnry) [reply docs/specviz2d/plugins.rst]: slightly modified your suggestion to '... by selecting "Manual" in the Background Type dropdown'

**Comment by ojustino** [comment]
> > When I choose the "Manual" option, I don't see a preview line and the 2D viewer takes 2-3 seconds to adjust when I change the center pixel.

> I've come across this a few times as well, but is unrelated to this PR (feel free to open an issue if one doesn't already exist).

I can see how the lag is unrelated, but do you have thoughts on the "Manual" type's lack of a live preview line and the error upon exporting a spectrum?

**Comment by kecnry** [comment]
> but do you have thoughts on the "Manual" type's lack of a live preview line and the error upon exporting a spectrum?

Addressed in #1737.  Good catch!

**Comment by kecnry** [comment]
> Would you consider adjusting the starting pixel for the trace when you select the "Manual" background type? With the image in the example notebook, it starts on the cross-dispersion pixel with the highest flux. 

Yes, this currently starts at the same default position that is determined for the initial trace, which probably isn't ideal.  I put a proposed improvement to that logic in #1738 (seemed like it should have its own change log entry since it isn't directly related to this PR, even though they're both dealing with the background subtraction step).

**Review by ojustino — APPROVED** [review]
Now that the tests are passing and the spinoff issues are underway, I say this branch is ready for a merge.

File: CHANGES.rst [hunk 3]
```diff
@@ -24,6 +24,10 @@ Specviz
 Specviz2d
 ^^^^^^^^^
 
+- Spectral extraction plugin now supports visualizing and exporting the 1D spectrum associated
+  with the background region. [#1682]
+
```
Comment (pllim) [inline CHANGES.rst:29]:
```suggestion
```

**Review by pllim — COMMENTED** [review]
Just a very minor nitpick. Code looks like code but should second approval come from a PO?

**Comment by pllim** [comment]
Ooops change log has conflict now.
